dodiscover/constraint/intervention.py

```diff
@@ -0,0 +1,362 @@
+                oriented_edges.append((node, nbr))
+        return added_arrows, oriented_edges
+
+    def _apply_rule12(
+        self, graph: EquivalenceClass, u: Column, a: Column, c: Column, context: Context
+    ) -> bool:
+        """Apply "Rule 9" of the I-FCI algorithm.
+
+        Checks for inducing paths where 'u' is the F-node, and 'a' and 'c' are connected:
+
+        'u' -> 'a' *-* 'c' with 'u' -> 'c', then orient 'a' -> 'c'.
+
+        For original details of the rule, see :footcite:`Kocaoglu2019characterization`.
+
+        Parameters
+        ----------
+        graph : EquivalenceClass
+            The causal graph.
+        u : Column
+            The candidate F-node
+        a : Column
+            Neighbors of the F-node.
+        c : Column
+            Neighbors of the F-node.
+        context : Context
+            The causal context.
+
+        Returns
+        -------
+        added_arrows : bool
+            Whether or not an orientation was made.
+
+        References
+        ----------
+        .. footbibliography::
+        """
+        f_nodes = context.f_nodes
+        symmetric_diff_map = context.symmetric_diff_map
+
+        added_arrows = False
+        if u in f_nodes and self.known_intervention_targets:
+            # get sigma map to map F-node to its symmetric difference target
+            S_set: FrozenSet = symmetric_diff_map.get(u, frozenset())
+
+            # check a *-* c
+            if (
+                len(S_set) == 1
+                and a in S_set
+                and (graph.has_edge(a, c) or graph.has_edge(c, a))
+                and graph.has_edge(u, a)
+                and graph.has_edge(u, c)
+            ):
+                # remove all edges between a and c
+                graph.remove_edge(a, c)
+                graph.remove_edge(c, a)
+
+                # then orient X -> Y
+                graph.add_edge(a, c, graph.directed_edge_name)
+
+                added_arrows = True
+        return added_arrows
+
+    def _apply_orientation_rules(self, graph: EquivalenceClass, sep_set: SeparatingSet):
+        idx = 0
+        finished = False
+
+        # apply R11, which is called R8 in I-FCI / Psi-FCI orienting all F-nodes
+        _ = self._apply_rule11(graph, self.context_)
+
+        while idx < self.max_iter and not finished:
+            change_flag = False
+            logger.info(f"Running R1-10 for iteration {idx}")
+
+            for u in graph.nodes:
+                for (a, c) in permutations(graph.neighbors(u), 2):
+                    logger.debug(f"Check {u} {a} {c}")
+
+                    # apply R1-3 to orient triples and arrowheads
+                    r1_add = self._apply_rule1(graph, u, a, c)
+                    r2_add = self._apply_rule2(graph, u, a, c)
+                    r3_add = self._apply_rule3(graph, u, a, c)
+
+                    # apply R4, orienting discriminating paths
+                    r4_add, _ = self._apply_rule4(graph, u, a, c, sep_set)
+
```

Comment:
I don't have the full context of the methods here. Rule 5, 6 and 7 do not apply here?

Reply:
Yes, these do not handle selection bias